lgc/builder/BuilderImpl.cpp

```diff
@@ -172,16 +173,63 @@ Value *BuilderImplBase::CreateIntegerDotProduct(Value *vector1, Value *vector2,
           indices = ArrayRef<int>({2, 3});
           input1 = CreateShuffleVector(input1, Constant::getNullValue(inputTy), indices);
           input2 = CreateShuffleVector(input2, Constant::getNullValue(inputTy), indices);
-          product =
+          scalar =
               CreateIntrinsic(intrinsicDot2, {}, {input1, input2, intermediateRes, getInt1(false)}, nullptr, instName);
         }
         // Add a satruation add if required.
-        if (!hasNoAccumulator)
-          scalar = CreateNamedCall("llvm.sadd.sat.i32", getInt32Ty(), {product, accumulator}, {}, instName);
+        if (hasAccumulator)
+          scalar = CreateNamedCall("llvm.sadd.sat.i32", getInt32Ty(), {scalar, accumulator}, {}, instName);
       }
     }
   }
 
+  // Do clamp if it has an accumulator
+  // NOTE: HW supported dot product will do clamp for 32-bit accumulator and no need to do it here
+  if ((!isHwSupported || (isHwSupported && scalar->getType() != outputTy))) {
```

Comment:
Niggle: redundant second isHwSupported, and too many parentheses for LLVM style.